pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionPool.java

```diff
@@ -439,17 +451,16 @@ public void close() throws Exception {
         }
     }
 
-    private void cleanupConnection(InetSocketAddress address, int connectionKey,
-            CompletableFuture<ClientCnx> connectionFuture) {
-        ConcurrentMap<Integer, CompletableFuture<ClientCnx>> map = pool.get(address);
-        if (map != null) {
-            map.remove(connectionKey, connectionFuture);
-        }
+    private void cleanupConnection(InetSocketAddress logicalAddress,
+                                   InetSocketAddress physicalAddress, int connectionKey,
```

Comment:
It seems that a better approach would be to pass the `Key` instance as a parameter.

Reply:
I'd also change the signature of `private CompletableFuture<ClientCnx> createConnection(InetSocketAddress logicalAddress, InetSocketAddress physicalAddress, int connectionKey)` to take the `Key` as a parameter. The public method should be left unchanged.

Reply:
> It seems that a better approach would be to pass the Key instance as a parameter.

I think we can remove this `cleanupConnection` and just directly call `pool.remove(key, connectionFuture);.` This `cleanupConnection` becomes a redundant one-line func if we accept the key.